File: rust/bbnf/src/grammar.rs
extern crate parse_that;

use parse_that::parsers::utils::escaped_span;
use parse_that::{
    any_span, lazy, next_span, string, string_span, take_while_span, Parser, ParserFlat,
    ParserSpan, ParserState, Span,
};

extern crate pretty;
use pretty::{Doc, Pretty};

use indexmap::IndexMap;

#[derive(Pretty, Debug, Clone, Copy, Eq, PartialEq)]
pub enum Comment<'a> {
    Line(&'a str),
    Block(&'a str),
}

#[derive(Pretty, Debug, Clone, Copy, Eq, PartialEq)]
pub struct Comments<'a> {
    pub left: Option<Comment<'a>>,
    pub right: Option<Comment<'a>>,
}

type TokenExpression<'a, T = Expression<'a>> = Box<Token<'a, T>>;

#[derive(Pretty, Debug, Clone, Eq, PartialEq, Hash)]
pub enum Expression<'a> {
    Literal(Token<'a, &'a str>),
    Nonterminal(Token<'a, &'a str>),

    Regex(Token<'a, &'a str>),

    MappingFn(Token<'a, String>),
    MappedExpression((TokenExpression<'a>, TokenExpression<'a>)),

    DebugExpression((TokenExpression<'a>, String)),

    Group(TokenExpression<'a>),
    Optional(TokenExpression<'a>),
    OptionalWhitespace(TokenExpression<'a>),

    Many(TokenExpression<'a>),
    Many1(TokenExpression<'a>),

    Skip(TokenExpression<'a>, TokenExpression<'a>),
    Next(TokenExpression<'a>, TokenExpression<'a>),
    Minus(TokenExpression<'a>, TokenExpression<'a>),

    Concatenation(TokenExpression<'a, Vec<Expression<'a>>>),
    Alternation(TokenExpression<'a, Vec<Expression<'a>>>),

    Rule(Box<Expression<'a>>, Option<Box<Expression<'a>>>),

    ProductionRule(Box<Expression<'a>>, Box<Expression<'a>>),

    Epsilon(Token<'a, ()>),
}

#[derive(Pretty, Debug, Clone, Copy, Eq)]
pub struct Token<'a, T> {
    pub value: T,

    #[pretty(skip)]
    pub span: Span<'a>,
    #[pretty(skip)]
    pub comments: Option<Comments<'a>>,
}

impl<'a, T> Token<'a, T> {
    pub fn new(value: T, span: Span<'a>) -> Self {
        Self {
            value,
            span,
            comments: None,
        }
    }

    pub fn new_without_span(value: T) -> Self {
        Self {
            value,
            span: Span::new(0, 0, ""),
            comments: None,
        }
    }
}

pub type AST<'a> = IndexMap<Expression<'a>, Expression<'a>>;

pub fn set_expression_comments<'a>(expr: &mut Expression<'a>, comments: Comments<'a>) {
    match expr {
        Expression::Literal(token) | Expression::Nonterminal(token) => {
            token.comments = Some(comments)
        }

        Expression::Regex(token) => token.comments = Some(comments),

        Expression::Epsilon(token) => token.comments = Some(comments),

        Expression::Group(token)
        | Expression::Optional(token)
        | Expression::Many(token)
        | Expression::Many1(token)
        | Expression::Skip(token, _)
        | Expression::Next(token, _)
        | Expression::Minus(token, _) => token.comments = Some(comments),

        Expression::Concatenation(token) | Expression::Alternation(token) => {
            token.comments = Some(comments)
        }

        _ => {}
    }
}

fn map_factor<'a>(
    factor: (Expression<'a>, Option<Span<'a>>),
    prev_offset: usize,
    state: &mut ParserState<'a>,
) -> Expression<'a> {
    match factor {
        (expr, Some(op)) => {
            let token = Token::new(expr, Span::new(prev_offset, state.offset, state.src));
            match op.as_str() {
                "*" => Expression::Many(Box::new(token)),
                "+" => Expression::Many1(Box::new(token)),
                "?w" => Expression::OptionalWhitespace(Box::new(token)),
                "?" => Expression::Optional(Box::new(token)),
                _ => unreachable!(
                    "unhandled factor: {:?}, {:?}",
                    op.as_str(),
                    token.span.as_str()
                ),
            }
        }
        (expr, _) => expr,
    }
}

fn reduce_binary_expression<'a>(
    expr: (Expression<'a>, Vec<(Span<'a>, Expression<'a>)>),
    prev_offset: usize,
    state: &mut ParserState<'a>,
) -> Expression<'a> {
    let (left, right) = expr;

    if right.is_empty() {
        return left;
    }

    right.into_iter().fold(left, |acc, (op, right)| {
        let acc_token = Token::new(acc, Span::new(prev_offset, state.offset, state.src));
        let right_token = Token::new(right, Span::new(prev_offset, state.offset, state.src));

        match op.as_str() {
            "<<" => Expression::Skip(Box::new(acc_token), Box::new(right_token)),
            ">>" => Expression::Next(Box::new(acc_token), Box::new(right_token)),
            "-" => Expression::Minus(Box::new(acc_token), Box::new(right_token)),
            _ => unreachable!(),
        }
    })
}

pub struct BBNFGrammar<'a> {
    _marker: std::marker::PhantomData<&'a ()>,
}

impl<'a> BBNFGrammar<'a> {
    fn block_comment() -> Parser<'a, Comment<'a>> {
        let not_comment = take_while_span(|c| c != '*' && c != '/');

        let comment = not_comment.many_span(1..);

        comment
            .wrap_span(string_span("/*"), string_span("*/"))
            .trim_whitespace()
            .many_span(1..)
            .map(|s| Comment::Block(s.as_str()))
    }

    fn line_comment() -> Parser<'a, Comment<'a>> {
        let not_newline = take_while_span(|c| c != '\n');
        let end = string_span("\r").opt_span().then_span(string_span("\n"));

        not_newline
            .wrap_span(string_span("//"), end)
            .many_span(1..)
            .map(|s| Comment::Line(s.as_str()))
    }

    fn identifier() -> Parser<'a, Span<'a>> {
        let first_part = take_while_span(|c| c.is_alphabetic() || c == '_');
        let rest_part =
            take_while_span(|c| c.is_alphanumeric() || c == '_' || c == '-' || c == '.')
                .many_span(..);
        first_part.then_span(rest_part)
    }

    fn literal() -> Parser<'a, Expression<'a>> {
        let quoted = |quote: &'a str| {
            let not_quote = take_while_span(|c| c != quote.chars().next().unwrap() && c != '\\');
            (not_quote | escaped_span())
                .many_span(..)
                .wrap_span(string_span(quote), string_span(quote))
        };

        (quoted("\"") | quoted("'") | quoted("`")).map(|s| {
            let token = Token::new(s.as_str(), s);
            Expression::Literal(token)
        })
    }

    fn epsilon() -> Parser<'a, Expression<'a>> {
        string_span("epsilon").map(|s| {
            let token = Token::new((), s);
            Expression::Epsilon(token)
        })
    }

    fn nonterminal() -> Parser<'a, Expression<'a>> {
        Self::identifier().map(|s| {
            let token = Token::new(s.as_str(), s);
            Expression::Nonterminal(token)
        })
    }

    fn regex() -> Parser<'a, Expression<'a>> {
        let not_slash = take_while_span(|c| c != '/');

        let escaped_span = string_span(r"\").then_span(string_span("/"));

        let string = (escaped_span | not_slash)
            .many_span(..)
            .wrap_span(string_span("/"), string_span("/"));

        string.map(|s| {
            match regex::Regex::new(s.as_str()) {
                Ok(_) => {}
                Err(e) => panic!("invalid regex: {:?}, {:?}", s.as_str(), e),
            }
            let token = Token::new(s.as_str(), s);
            Expression::Regex(token)
        })
    }

    fn group() -> Parser<'a, Expression<'a>> {
        lazy(|| {
            Self::rhs()
                .trim_whitespace()
                .wrap(string_span("("), string_span(")"))
                .map_with_state(|expr, prev_offset, state| {
                    let token = Token::new(expr, Span::new(prev_offset, state.offset, state.src));
                    Expression::Group(Box::new(token))
                })
        })
    }

    fn optional_group() -> Parser<'a, Expression<'a>> {
        lazy(|| {
            Self::rhs()
                .trim_whitespace()
                .wrap(string_span("["), string_span("]"))
                .map_with_state(|expr, prev_offset, state| {
                    let token = Token::new(expr, Span::new(prev_offset, state.offset, state.src));
                    Expression::Optional(Box::new(token))
                })
        })
    }

    fn many_group() -> Parser<'a, Expression<'a>> {
        lazy(|| {
            Self::rhs()
                .trim_whitespace()
                .wrap(string_span("{"), string_span("}"))
                .map_with_state(|expr, prev_offset, state| {
                    let token = Token::new(expr, Span::new(prev_offset, state.offset, state.src));
                    Expression::Many(Box::new(token))
                })
        })
    }

    fn term() -> Parser<'a, Expression<'a>> {
        Self::epsilon()
            | Self::group()
            | Self::optional_group()
            | Self::many_group()
            | Self::nonterminal()
            | Self::literal()
            | Self::regex()
    }

    fn trim_comment(
        p: Parser<'a, Expression<'a>>,
        comment_parser: Parser<'a, Option<Comment<'a>>>,
    ) -> Parser<'a, Expression<'a>> {
        p.trim_keep(comment_parser).map(|(left, mut expr, right)| {
            if left.is_some() || right.is_some() {
                let comments = Comments { left, right };
                set_expression_comments(&mut expr, comments);
            }
            expr
        })
    }

    fn factor() -> Parser<'a, Expression<'a>> {
        Self::trim_comment(
            Self::term()
                .then(any_span(&["?w", "*", "+", "?"]).trim_whitespace().opt())
                .map_with_state(map_factor),
            Self::block_comment().opt(),
        )
    }

    fn binary_factor() -> Parser<'a, Expression<'a>> {
        Self::factor()
            .then(
                any_span(&["<<", ">>", "-"])
                    .trim_whitespace()
                    .then(Self::factor())
                    .many(..),
            )
            .map_with_state(reduce_binary_expression)
    }

    fn concatenation() -> Parser<'a, Expression<'a>> {
        let delim = string_span(",").trim_whitespace();

        Self::binary_factor()
            .sep_by(delim, ..)
            .map_with_state(|exprs, prev_offset, state| {
                if exprs.len() == 1 {
                    exprs.into_iter().next().unwrap()
                } else {
                    let token = Token::new(exprs, Span::new(prev_offset, state.offset, state.src));
                    Expression::Concatenation(Box::new(token))
                }
            })
    }

    fn alternation() -> Parser<'a, Expression<'a>> {
        let delim = string_span("|").trim_whitespace();

        Self::concatenation()
            .sep_by(delim, ..)
            .map_with_state(|exprs, prev_offset, state| {
                if exprs.len() == 1 {
                    exprs.into_iter().next().unwrap()
                } else {
                    let token = Token::new(exprs, Span::new(prev_offset, state.offset, state.src));
                    Expression::Alternation(Box::new(token))
                }
            })
    }

    fn lhs() -> Parser<'a, Expression<'a>> {
        Self::nonterminal()
    }

    fn rhs() -> Parser<'a, Expression<'a>> {
        Self::alternation()
    }

    fn mapping_fn() -> Parser<'a, Option<Box<Expression<'a>>>> {
        let lhs = string_span(";").skip(Self::lhs().trim_whitespace().skip(string_span("=")));
        let not_lhs = next_span(1).look_ahead(lhs.negate());

        string_span("=>")
            .trim_whitespace()
            .next(not_lhs.many_span(..).then_span(next_span(1)))
            .map(|s| {
                let token = Token::new(s.as_str().to_string(), s);
                match syn::parse_str::<syn::ExprClosure>(s.as_str()) {
                    Ok(_) => {}
                    Err(e) => panic!("invalid mapper expression: {:?}, {:?}", s.as_str(), e),
                }

                Box::new(Expression::MappingFn(token))
            })
            .opt()
    }

    fn production_rule() -> Parser<'a, Expression<'a>> {
        let comment = Self::block_comment() | Self::line_comment();
        let eq = string("=").trim_whitespace();

        let terminator = (any_span(&[";", "."])).trim_whitespace();

        let production_rule = Self::lhs()
            .skip(eq)
            .then(Self::rhs())
            .then_flat(Self::mapping_fn())
            .skip(terminator)
            .map(|(lhs, rhs, mapping_fn)| {
                Expression::ProductionRule(
                    lhs.into(),
                    Expression::Rule(Box::new(rhs), mapping_fn).into(),
                )
            });

        Self::trim_comment(production_rule, comment.opt())
    }

    pub fn grammar() -> Parser<'a, AST<'a>> {
        let rule = Self::production_rule().trim_whitespace().many(..);

        return rule.trim_whitespace().map(|rules| {
            rules
                .into_iter()
                .map(|expr| match expr {
                    Expression::ProductionRule(lhs, rhs) => (*lhs, *rhs),
                    _ => unreachable!(),
                })
                .collect()
        });
    }
}

impl<'a, T: PartialEq> PartialEq for Token<'a, T> {
    fn eq(&self, other: &Self) -> bool {
        self.value == other.value
    }
}
impl<'a, T: std::hash::Hash> std::hash::Hash for Token<'a, T> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.value.hash(state);
    }
}

File: typescript/src/bbnf/grammar.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import { Parser, string, all, any, regex } from "../parse/index.js";
import type { Expression, Literal, Epsilon, Nonterminal, Comment, Regex, Group, Optional, Many, Concatenation, Alteration, ProductionRule } from "./types.js";

const operatorToType: Record<string, string> = {
    "|": "alternation",
    ",": "concatenation",
    "-": "minus",
    "<<": "skip",
    ">>": "next",
    "*": "many",
    "+": "many1",
    "?": "optional",
    "?w": "optionalWhitespace",
};

const reduceBinaryExpression = ([left, rightExpression]: [any, any[]]) => {
    if (rightExpression.length === 0) {
        return left;
    }
    return rightExpression.reduce((acc: Expression, [op, right]: [string, Expression]) => {
        return {
            type: operatorToType[op],
            value: [acc, right],
        } as Expression;
    }, left);
};

const mapFactor = ([term, op]: [any, any]) => {
    if (op === undefined) {
        return term;
    }
    const type = operatorToType[op];
    return {
        type,
        value: term,
    } as Expression;
};

function mapStatePosition(parser: Parser<any>) {
    return parser.mapState((newState, oldState) => {
        if (newState.value && newState.value.range === undefined) {
            newState.value.range = {
                start: oldState.offset,
                end: newState.offset,
            };
        }
        return newState;
    });
}

type Options = {
    debug: boolean;
    comments: boolean;
};

const defaultOptions: Options = {
    debug: false,
    comments: true,
};

export class BBNFGrammar {
    options: Options;

    // Backing fields for lazy-memoized parsers
    private _blockComment?: Parser<any>;
    private _lineComment?: Parser<any>;
    private _comment?: Parser<any>;
    private _group?: Parser<any>;
    private _regex?: Parser<any>;
    private _optionalGroup?: Parser<any>;
    private _manyGroup?: Parser<any>;
    private _lhs?: Parser<any>;
    private _term?: Parser<any>;
    private _factor?: Parser<any>;
    private _binaryFactor?: Parser<any>;
    private _concatenation?: Parser<any>;
    private _alternation?: Parser<any>;
    private _rhs?: Parser<any>;
    private _productionRule?: Parser<any>;
    private _grammar?: Parser<any>;

    constructor(options?: Partial<Options>) {
        this.options = {
            ...defaultOptions,
            ...(options ?? {}),
        };
    }

    identifier() {
        return regex(/[_a-zA-Z][_a-zA-Z0-9-]*/);
    }

    literal() {
        return any(
            regex(/(\\.|[^"\\])*/).wrap(string('"'), string('"')),
            regex(/(\\.|[^'\\])*/).wrap(string("'"), string("'")),
            regex(/(\\.|[^`\\])*/).wrap(string("`"), string("`")),
        ).map((value: any) => {
            value = value.replace(/\\(.)/g, "$1");
            return {
                type: "literal",
                value,
            } as Literal;
        });
    }

    epsilon() {
        return any(string("epsilon"), string("ε")).map(() => {
            return {
                type: "epsilon",
            } as Epsilon;
        });
    }

    nonterminal() {
        return this.identifier().map((value) => {
            return {
                type: "nonterminal",
                value,
            } as Nonterminal;
        });
    }

    blockComment(): Parser<any> {
        return (this._blockComment ??= Parser.lazy(() =>
            mapStatePosition(
                regex(/\/\*[^\*]*\*\//).map((v) => {
                    return {
                        type: "comment",
                        value: v,
                    } as Comment;
                }),
            ),
        ));
    }

    lineComment(): Parser<any> {
        return (this._lineComment ??= Parser.lazy(() =>
            mapStatePosition(
                regex(/\/\/.*/).map((v) => {
                    return {
                        type: "comment",
                        value: v,
                    } as Comment;
                }),
            ),
        ));
    }

    comment(): Parser<any> {
        return (this._comment ??= Parser.lazy(() =>
            any(this.blockComment(), this.lineComment()),
        ));
    }

    trimBigComment(e: Parser<any>) {
        return e
            .trim(this.blockComment().trim().many() as any, false)
            .map(([left, expression, right]: any) => {
                expression.comment = {
                    left,
                    right,
                };
                return expression as unknown as Expression;
            }) as Parser<Expression>;
    }

    group(): Parser<any> {
        return (this._group ??= Parser.lazy(() =>
            this.rhs()
                .trim()
                .wrap(string("("), string(")"))
                .map((value: any) => {
                    return {
                        type: "group",
                        value,
                    } as Group;
                }),
        ));
    }

    regexRule(): Parser<any> {
        return (this._regex ??= Parser.lazy(() =>
            regex(/(\\.|[^\/])+/)
                .wrap(string("/"), string("/"))
                .then(regex(/[gimuy]*/).opt())
                .map(([r, flags]: any) => {
                    return {
                        type: "regex",
                        value: new RegExp(r, flags ?? undefined),
                    } as Regex;
                }),
        ));
    }

    optionalGroup(): Parser<any> {
        return (this._optionalGroup ??= Parser.lazy(() =>
            this.rhs()
                .trim()
                .wrap(string("["), string("]"))
                .map((value: any) => {
                    return {
                        type: "optional",
                        value: {
                            type: "group",
                            value,
                        },
                    } as Optional;
                }),
        ));
    }

    manyGroup(): Parser<any> {
        return (this._manyGroup ??= Parser.lazy(() =>
            this.rhs()
                .trim()
                .wrap(string("{"), string("}"))
                .map((value: any) => {
                    return {
                        type: "many",
                        value: {
                            type: "group",
                            value,
                        },
                    } as Many;
                }),
        ));
    }

    lhs(): Parser<any> {
        return (this._lhs ??= Parser.lazy(() =>
            mapStatePosition(this.nonterminal()),
        ));
    }

    term(): Parser<any> {
        return (this._term ??= Parser.lazy(() =>
            mapStatePosition(
                any(
                    this.epsilon(),
                    this.group(),
                    this.optionalGroup(),
                    this.manyGroup(),
                    this.nonterminal(),
                    this.literal(),
                    this.regexRule(),
                ) as any,
            ),
        ));
    }

    factor(): Parser<any> {
        return (this._factor ??= Parser.lazy(() =>
            this.trimBigComment(
                mapStatePosition(
                    all(
                        this.term(),
                        any(
                            string("?w"),
                            string("?"),
                            string("*"),
                            string("+"),
                        )
                            .trim()
                            .opt() as any,
                    ).map(mapFactor),
                ),
            ),
        ));
    }

    binaryFactor(): Parser<any> {
        return (this._binaryFactor ??= Parser.lazy(() =>
            mapStatePosition(
                all(
                    this.factor(),
                    all(
                        any(string("<<"), string(">>"), string("-")).trim(),
                        this.factor(),
                    ).many(),
                ).map(reduceBinaryExpression),
            ),
        ));
    }

    concatenation(): Parser<any> {
        return (this._concatenation ??= Parser.lazy(() =>
            mapStatePosition(
                this.binaryFactor().sepBy(string(",").trim()),
            ).map((value: any) => {
                if (value.length === 1) {
                    return value[0];
                }

                return {
                    type: "concatenation",
                    value,
                } as Concatenation;
            }),
        ));
    }

    alternation(): Parser<any> {
        return (this._alternation ??= Parser.lazy(() =>
            mapStatePosition(
                this.concatenation().sepBy(string("|").trim()),
            ).map((value: any) => {
                if (value.length === 1) {
                    return value[0];
                }

                return {
                    type: "alternation",
                    value,
                } as Alteration;
            }),
        ));
    }

    rhs(): Parser<any> {
        return (this._rhs ??= Parser.lazy(() =>
            this.alternation(),
        ));
    }

    productionRule(): Parser<any> {
        return (this._productionRule ??= Parser.lazy(() =>
            all(
                this.lhs(),
                string("=").trim(),
                this.rhs(),
                any(string(";"), string(".")).trim(),
            ).map(([name, , expression]: any) => {
                return { name, expression } as ProductionRule;
            }),
        ));
    }

    grammar(): Parser<any> {
        return (this._grammar ??= Parser.lazy(() =>
            this.productionRule()
                .trim(this.lineComment().trim().many() as any, false)
                .map(([above, rule, below]: any) => {
                    rule.comment = {
                        above,
                        below,
                    };
                    return rule;
                })
                .many(1)
                .trim(),
        ));
    }
}
